R/transformAbundance.R

```diff
@@ -0,0 +1,400 @@
+                  Z = .get_z_table,
+                  log10 = .get_log10_table,
+                  log10p = .get_log10p_table,
+                  hellinger = .get_hellinger_table,
+                  identity = .get_identity_table,
+                  clr = .get_clr_table)
+
+    # Does the function call, arguments are "assay" abundance table and "pseudocount"
+    do.call(FUN,
+            list(assay = assay,
+                 target = target))
+}
+
+.get_relabundance_table <- function(assay, target){
+
+    # Calculates the relative abundances
+    mat <- sweep(assay, 2, colSums(assay), "/")
+    return(mat)
+}
+
+.get_z_table <- function(assay, target){
+
+    # Log10 can not be calculated if there is zero
+    if (any(assay == 0)) {
+        stop("Abundance table contains zero and Z transformation
+                is being applied without pseudocount. Try Z with
+                pseudocount=TRUE.")
+    }
+
+    # Gets the log transformed table
+    mat <- .get_log10_table(assay, target)
+
+    if(target=="features"){
+        # Z transform for features. Centers the feature data. After that, divides with
+        # the standard deviation of feature.
+        trans <- t(scale(t(mat)))
+        # Saves all features that are undetectable in every sample i.e. are NA
+        undetectables <- which(rowMeans(is.na(trans)) == 1)
+
+        # If there are some deatures that are undetectable
+        # i.e. "undetectable" has length over 0, and table have zeros
+        if (length(undetectables) > 0 & min(mat) == 0) {
+
+            warning("Some features were not detectable. In all samples, signal was
+                    under detectable limit.")
+
+            # Some features are undetectable in all samples, they are NA when scaled.
+            # 0 is assigned to those features.
+
+            trans[names(undetectables), ] <- 0
+
+            mat <- trans
+
+            # Deletes extra information
+            attr(mat,"scaled:scale") <- NULL
+            attr(mat,"scaled:center") <- NULL
+        }
+
+    }else if(target=="samples"){
+        # Performs z transformation for samples
+        mat <- apply(mat, 2, function(x) {
+            (x - mean(x))/sd(x)
+        })
+    }
+
+    return(mat)
+}
+
+.get_log10_table <- function(assay, target){
+
+    # If abundance table contains zeros, gives an error, because it is not possible
+    # to calculate log from zeros. If there is no zeros, calculates log.
+    if (any(assay == 0)) {
+        stop("Abundance table contains zero and log10 transformation
+                is being applied without pseudocount. Try log10 with
+                pseudocount=TRUE or log10p transformation.")
+    } else {
+        mat <- log10(assay)
+    }
+    return(mat)
+}
+
+.get_log10p_table <- function(assay, target){
+
+    # Because a pseudo count was added, there is no zeroes in the abundance table.
+    # Calculates log "directly".
+    mat <- log10(assay)
+
+    return(mat)
+}
```

Comment:
Have a look at `normalizeCounts` in the `scuttle` package. I think you should reuse the implementation from there. No need to reinvent the wheel

Reply:
I have thought this a while... Do you mean log1p()? I can not find any other helpful things from the code. If so, that does the natural logarithm

Reply:
We like to do log10 in microbiome data, instead of natural log. But does it work with that change?

Reply:
I am just wondering, what is the functionality that is wanted to be implemented also here. 

Well, my opinion might be very biased, but I think our code is quite straight-forward. All in all, log10 and log10p are the same. Just the pseudocount-adding differs there, and I think it is done in quite logical point, which can be easily found from the code. 

Also, user have choice to choose if the pseudocount is 2 and not 1 like it is by default in log10p.

Reply:
True. Thinking about it now, I do not think that we need both "log10" and "log10p" for this reason. We can just keep the "log10" with the option of pseudocount. I think the original idea had been to have a "quick" shortcut for log10 + pseudocount=1 but I guess it is potentially confusing and "log10p" option can be removed.

Reply:
log10p removed